uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/annual-entitlement-change-log.component.js

```diff
@@ -0,0 +1,319 @@
+/* eslint-env amd */
+
+define([
+  'common/lodash',
+  'common/moment',
+  'leave-absences/absence-tab/modules/components',
+  'common/models/contract'
+], function (_, moment, components) {
+  components.component('annualEntitlementChangeLog', {
+    bindings: {
+      contactId: '<',
+      periodId: '<',
+      dismissModal: '&'
+    },
+    templateUrl: ['settings', function (settings) {
+      return settings.pathTpl + 'components/annual-entitlement-change-log.html';
+    }],
+    controllerAs: 'changeLog',
+    controller: AnnualEntitlementChangeLog
+  });
+
+  AnnualEntitlementChangeLog.$inject = ['AbsencePeriod', 'AbsenceType',
+    'Entitlement', 'OptionGroup', 'shared-settings'];
+
+  function AnnualEntitlementChangeLog (AbsencePeriod, AbsenceType, Entitlement,
+  OptionGroup, sharedSettings) {
+    var entitlementsChangeLog = [];
+    var vm = this;
+
+    vm.absencePeriod = null;
+    vm.absenceTypes = [];
+    vm.changeLogRows = [];
+    vm.loading = { component: true };
+
+    (function init () {
+      loadAbsencePeriod()
+      .then(loadCalculationUnits)
+      .then(loadAbsenceTypes)
+      .then(loadChangeLog)
+      .then(appendCurrentEntitlementsToChangeLog)
+      .then(createChangeLogRows)
+      .then(removeRepeatedComments)
+      .then(highlightEntitlementWithComments)
+      .finally(function () {
+        vm.loading.component = false;
+      });
+    })();
+
+    /**
+     * Appends the current entitlements for the user and period into the change
+     * log's list. This is done because the logs don't return the current
+     * entitlement values, just the previous ones.
+     *
+     * @return {Promise}
+     */
+    function appendCurrentEntitlementsToChangeLog () {
+      return getCurrentEntitlementsLog().then(function (currentEntitlements) {
+        entitlementsChangeLog = entitlementsChangeLog.concat(currentEntitlements);
+      });
+    }
+
+    /**
+     * Creates the structure used to display entitlement logs by grouping
+     * changes using their creation date and sorting them in a descending order.
+     */
+    function createChangeLogRows () {
+      var groupedEntitlements = _.groupBy(entitlementsChangeLog,
+        groupByCreationDateRoundedSeconds);
+
+      vm.changeLogRows = _.map(groupedEntitlements, getChangeLogRow)
+        .sort(function (previousRow, currentRow) {
+          return currentRow.date.diff(previousRow.date);
+        });
+    }
+
+    /**
+     * Selects the entitlement rows to be highlighlighted based on the one
+     * that has comments.
+     */
+    function highlightEntitlementWithComments () {
+      for (var i = vm.changeLogRows.length - 1; i >= 0; i--) {
+        var changeLogRow = vm.changeLogRows[i];
```

Comment:
All `vars` should be defined earlier before the cycle, right at the start of the function.